### ISSUE TITLE
TypePath required for generic #[reflect(ignore)] field

### ISSUE THREAD

**Issue by paul-hansen** [bevyengine/bevy#9094]
## Bevy version

Bevy 0.11

## What you did

I'm working on a PR to update leafwing-input-manager: https://github.com/Leafwing-Studios/leafwing-input-manager/pull/364

## What went wrong
Leafwing input manager has a struct ActionState that takes a generic that is used on a `#[reflect(ignore)]` PhantomData field:
```rust
#[derive(Resource, Component, Clone, Debug, PartialEq, Serialize, Deserialize, Reflect)]
pub struct ActionState<A: Actionlike> {
    /// The [`ActionData`] of each action
    ///
    /// The position in this vector corresponds to [`Actionlike::index`].
    action_data: Vec<ActionData>,
    #[reflect(ignore)]
    _phantom: PhantomData<A>,
}
```

We are getting an error when registering this type with bevy saying our generic type must implement TypePath.
This seems potentially limiting to the end users of LWIM and I'm not sure it was intended as it isn't mentioned in migration guides.
Is this restriction something that could be lifted?

## Additional info
There's a conversation about this on the LWIM PR starting here: 
https://github.com/Leafwing-Studios/leafwing-input-manager/pull/364#issuecomment-1628136788

To try this out, you can checkout the branch for that PR and remove the TypePath requirement I added from this line:
https://github.com/Leafwing-Studios/leafwing-input-manager/blob/7fb55b001a4aa949b1863dd2e0007d9512960625/src/lib.rs#L81

Here's the line and error we are getting:
https://github.com/Leafwing-Studios/leafwing-input-manager/blob/dfbe44b22fae01434737fb498dfb10d5ca20057d/src/plugin.rs#L146
```rust
        app.register_type::<ActionState<A>>()
```

```
error[E0277]: the trait bound `A: TypePath` is not satisfied
   --> src\plugin.rs:146:29
    |
146 |         app.register_type::<ActionState<A>>()
    |                             ^^^^^^^^^^^^^^ the trait `TypePath` is not implemented for `A`
    |
note: required for `ActionState<A>` to implement `GetTypeRegistration`
   --> src\action_state.rs:85:80
    |
85  | #[derive(Resource, Component, Clone, Debug, PartialEq, Serialize, Deserialize, Reflect)]
    |                                                                                ^^^^^^^ unsatisfied trait bound introduced in this `derive` macro
86  | pub struct ActionState<A: Actionlike> {
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^
note: required by a bound in `App::register_type`
   --> D:\Packages\.cargo\registry\src\index.crates.io-6f17d22bba15001f\bevy_app-0.11.0\src\app.rs:816:29
    |
816 |     pub fn register_type<T: bevy_reflect::GetTypeRegistration>(&mut self) -> &mut Self {
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `App::register_type`
    = note: this error originates in the derive macro `Reflect` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting this bound
    |
82  | impl<A: Actionlike + bevy::reflect::TypePath> Plugin for InputManagerPlugin<A> {
    |                    +++++++++++++++++++++++++
```

**Comment by MrGVSV** [comment]
Unfortunately, this cannot be lifted. The new, stable type paths require that all generics also implement `TypePath`. This is so that we can effectively create the necessary `TypePath` impl for the type. Without it, serialization and other type-name-specific workflows are not guaranteed to be stable between builds.

The solution would probably be for LWIM to add `Actionlike: TypePath` or to just ensure that whatever `A` is passed implements `TypePath`.

**Comment by paul-hansen** [comment]
That's what I figured, thought I'd ask though just in case. Thanks for the confirmation! I doubt this will be an issue for any LWIM users since they will likely just be using their own enum, they can just derive reflect on if they haven't already. 

Hopefully cases where crates don't own either type can find a work around, I'd guess make a wrapper struct and manually implement TypePath? Is this something worth mentioning in the migration guide? https://bevyengine.org/learn/migration-guides/0.10-0.11/#bevy-reflect-stable-type-path-v2

**Comment by paul-hansen** [comment]
Created a PR for the migration guide, don't think there's anything else to do here so I'll close this. I welcome feedback on that PR: https://github.com/bevyengine/bevy-website/pull/698

**Comment by MrGVSV** [comment]
I'm actually going to re-open this issue because I think we may want to consider adding alternative options for when types cannot possibly implement `TypePath`, as in the case with [`bevy_rand`](https://discord.com/channels/691052431525675048/1002362493634629796/1127868615996231780).

For such scenarios, we have a couple options:
1. Require these types to take in wrappers over other types
    ```rust
    #[derive(Reflect)]    
    struct Wrapper(external_crate::NotReflect);

    #[derive(Reflect)]
    struct MyType<T>(T);

    app.register_type::<MyType<Wrapper>>();
    ```
2. Add an opt-out attribute like `#[type_path(unstable)]` which implements all of the `TypePath` methods using `std::any::type_name`
3. Add an attribute to allow individual type parameters to be excluded from the type path: `#[type_path(ignore_params(T))]`

1 is annoying but currently possible. 

I think 2 is feasible, but would require more work to properly generate the `TypePath::short_name` impl. I think it also makes sense to use for types that really don't need a stable type path (i.e. there's no expectation they will be the same between builds).

I personally don't think 3 is the way to go, but it's an option.

**Comment by Bluefinger** [comment]
Option 3 won't really work for types that are meant to be generic and not clash, so that information has to be present somehow. Option 2 seems like the 'safest' bet, but as long as it can be 'contained' to just the part that has been opted out. I don't know whether for the case of `bevy_rand` that just keeping the opt-out on the generic parameter would suffice for a 'good enough' type path that is mostly stable (or however stable we can guarantee it)

**Comment by simnaut** [comment]
**TL;DR:** Foreign marker types used as generic parameters in reflected components hit a fourth shape that none of MrGVSV's three options fully cover. A sealed `MarkerType` trait with a blanket `TypePath` impl might be the missing piece, ideally gated on a `bevy_reflect_traits` extraction.

**Concrete repro (typed-quantity / phantom-frame domain crate):**

I'm hitting this in a Rust port of NASA JEOD where physics lives in a Bevy-free `jeod_quantities` crate (reused outside Bevy: a non-Bevy runner, batch propagation, future WASM):

```rust
// jeod_quantities (Bevy-free, no bevy_reflect dep)
pub trait Planet: 'static {}
pub struct Earth;
impl Planet for Earth {}
pub struct PlanetInertial<P: Planet>(PhantomData<P>);

// bevy_jeod (adapter)
#[derive(Component, Reflect)]
pub struct TranslationalStateC<P: Planet>(pub Position<PlanetInertial<P>>);

// Trying to bridge:
impl bevy::reflect::TypePath for jeod_quantities::Earth { /* ... */ }
//   ^^^ E0210: orphan rule — both `TypePath` and `Earth` are foreign.
```

Same shape surfaces in `uom`, `dimensioned`, and any phantom-typed physics / astro / units crate.

**How the three options map:**

- **Option 1 (newtype wrap):** doubles marker surface; loses `Position<PlanetInertial<Earth>>` ergonomics. Scales poorly past a fixed set.
- **Option 2 (`#[type_path(unstable)]`):** sacrifices stable scene round-trip for the *whole* struct, not just the phantom.
- **Option 3 (`#[type_path(ignore_params(T))]`):** closest fit, but collapses `TranslationalStateC<Earth>` and `<Mars>` into one registry key. Breaks when the phantom is identity-bearing (mine is — they're distinct components on different entities).

All three complement rather than fully answer the ask: *the consumer opts the foreign type into `TypePath` from the consumer crate, without the foreign crate cooperating, and without collapsing the registry.*

**Possible fourth shape:**

Sealed `pub trait MarkerType: 'static {}` in `bevy_reflect` with `impl<T: MarkerType> TypePath for T` returning `core::any::type_name::<T>()`. Consumer writes `impl bevy_reflect::MarkerType for Earth {}` — orphan-legal, no parameter collapse. Tradeoff: marker `type_path()` is `type_name`-stable, which Bevy rejected for stable scenes generally — but for a phantom that's never directly deserialized, this might be the right tradeoff. Sits next to options 1–3, doesn't subsume them.

Works cleanest if `TypePath` lives in a `bevy_reflect_traits` sub-crate — see [#22721](https://github.com/bevyengine/bevy/issues/22721) (ashivaram23's separation question) and [#20337](https://github.com/bevyengine/bevy/issues/20337) (umbrella). Happy to file separately or fold into [#22721](https://github.com/bevyengine/bevy/issues/22721), whichever lands cleaner.
